include/commissioner/network_diag_data.hpp

```diff
@@ -0,0 +1,187 @@
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file defines the types of Thread Network Diagnostic TLVs used for network diagnostics.
+ */
+
+#ifndef OT_COMM_NETWORK_DIAG_TLVS_HPP_
+#define OT_COMM_NETWORK_DIAG_TLVS_HPP_
+
+#include <cstddef>
+#include <cstdint>
+#include <stdbool.h>
+#include <string>
+#include <vector>
+
+#include "defines.hpp"
+#include "error.hpp"
+
+#define CHILD_TABLE_ENTRY_BYTES 3
+#define IPV6_ADDRESS_BYTES 16
+#define LEADER_DATA_BYTES 8
+#define MAC_COUNTERS_BYTES 36
+#define RLOC16_BYTES 2
+#define ROUTER_ID_MASK_BYTES 8
+
+namespace ot {
+
+namespace commissioner {
+
+/**
+ * @brief Mode Data
+ */
+struct ModeData
+{
+    bool mIsRxOnWhenIdleMode          = false;
+    bool mIsMtd                       = false;
+    bool mIsStableNetworkDataRequired = false;
+};
+
+/**
+ * @brief Child Entry in Child Table
+ */
+struct ChildTableEntry
+{
+    uint32_t mTimeout             = 0;
+    uint8_t  mIncomingLinkQuality = 0;
+    uint16_t mChildId             = 0;
+    ModeData mModeData;
+};
+
+/**
+ * @brief IPv6 Addresses List
+ */
+struct Ipv6AddressList
+{
+    std::vector<std::string> mIpv6Addresses;
+};
+
+/**
+ * @brief Leader Data
+ */
+struct LeaderData
+{
+    uint32_t mPartitionId       = 0;
+    uint8_t  mWeighting         = 0;
+    uint8_t  mDataVersion       = 0;
+    uint8_t  mStableDataVersion = 0;
+    uint8_t  mRouterId          = 0;
+};
+
+/**
+ * @brief Route Data Entry of RouteData in Route64
+ */
+struct RouteDataEntry
+{
+    uint8_t mRouterId            = 0;
+    uint8_t mOutgoingLinkQuality = 0;
+    uint8_t mIncomingLinkQuality = 0;
+    uint8_t mRouteCost           = 0;
+};
+
+/**
+ * @brief Route64 Data
+ */
+struct Route64
+{
+    uint8_t                     mIdSequence = 0;
+    ByteArray                   mMask;
+    std::vector<RouteDataEntry> mRouteData;
+};
+
+/**
+ * @brief Child IPv6 Address List
+ */
+struct ChildIpv6AddressList
+{
+    uint16_t        mRloc16 = 0;
+    Ipv6AddressList mIpv6AddressList;
```

Comment:
How can we identify which Child has those address? Should there be a `Child ID` or something similar?

Reply:
we can identify the child with rloc16

Reply:
Them how can we associate those addresses with a ChildTableEntry? There doesn't seem to be a rloc16 in ChildTableEntry...

Reply:
Although we can use `rloc16 & 0x1FF` to generate the `child ID` which can match with `ChildTableEntry `'s address, in practical applications, we still need to use RLOC16 to identify nodes in the network. In other words, the child ID in the Child Table Entry and the router ID of the device itself are used to generate the RLOC16 of that child.

Reply:
This means the client app developers need to know that they need to get the child id from `rloc16 & 0x1FF`, this is a knowledge barrier. Can we include also the `ChildId` in the `ChildIpv6AddressInfo` class? Just as there is a `ChildId` in `ChildTableEntry`. Or provide a member method `getChildId() const {...}`

Reply:
Added a new field named `ChildId` into the `ChildTableEntry`.